### PR TITLE
added support of "around" hooks of controllers

### PR DESCRIPTION
**Pull request by abolabo**

### CODE REVIEW

File: public_html/core/engine/dispatcher.php [hunk 1]
```diff
@@ -286,86 +271,95 @@ public function dispatch($parent_controller = '')
...
+                //clean up and destroy the object
+                unset($controller, $dispatch);
+            } else {
+                $err = new AError(
+                    'Error: controller method not exist '.$this->class.'::'.$this->method.'!',
+                    AC_ERR_CLASS_METHOD_NOT_EXIST
+                );
+                $err->toLog()->toDebug();
+                if (in_array($this->controller_type, ['responses', 'api', 'task'])) {
+                    $dd = new ADispatcher('responses/error/ajaxerror/not_found');
```
Comment (abantecart) [inline public_html/core/engine/dispatcher.php:355]: Why ajax error?

File: public_html/core/engine/dispatcher.php [hunk 2]
```diff
@@ -286,86 +271,95 @@ public function dispatch($parent_controller = '')
...
+                    $dd->dispatch();
+                }
+            }
+        } //catching output of around hook (it can be only one)
+        catch (AException $e) {
+            if ($e->getCode() != AC_HOOK_AROUND) {
```
Comment (abantecart) [inline public_html/core/engine/dispatcher.php:361]: If this functionality takes entire control of the execution, it should be called takeover or bypass hook. We need to rename it in all places.

File: public_html/core/lib/exceptions/exception_codes.php [hunk 3]
```diff
@@ -18,14 +18,8 @@
...
 
-if (!defined('E_DEPRECATED')) {
-    define('E_DEPRECATED', 8192);
-} // since PHP 5.3.0
-if (!defined('E_RECOVERABLE_ERROR')) {
-    define('E_RECOVERABLE_ERROR', 4096);
-} // since PHP 5.2.0
-
 // Error codes
+define('AC_HOOK_AROUND', 9999);
```
Comment (abantecart) [inline public_html/core/lib/exceptions/exception_codes.php:22]: Please rename

File: public_html/core/engine/dispatcher.php [hunk 4]
```diff
@@ -286,86 +271,95 @@ public function dispatch($parent_controller = '')
...
-            if ($dispatch && is_object($dispatch)) {
-                if ($this->args["instance_id"] == 0) {
-                    //If main controller come back for new dispatch
-                    return $dispatch->getController().'/'.$dispatch->getMethod();
+        try {
+            if (is_callable([$controller, $this->method])) {
+                /**
+                 * @var $dispatch ADispatcher
+                 */
+                $r = new ReflectionMethod($controller, $this->method);
```
Comment (abantecart) [inline public_html/core/engine/dispatcher.php:279]: What is ReflectionMethod?

File: public_html/core/engine/dispatcher.php [hunk 5]
```diff
@@ -286,86 +271,95 @@ public function dispatch($parent_controller = '')
...
+
+                $dispatch = call_user_func_array([$controller, $this->method], $args);
+                //Check if return is a dispatch and need to call new page
+                if ($dispatch && is_object($dispatch)) {
+                    if ($this->args["instance_id"] == 0) {
+                        //If main controller come back for new dispatch
+                        return $dispatch->getController().'/'.$dispatch->getMethod();
+                    } else {
+                        // Call new dispatch for new controller and exit
+                        //???? need to put limit for recursion to prevent overflow
```
Comment (abantecart) [inline public_html/core/engine/dispatcher.php:295]: Should we address this issue? If not, we need to remove the comment

**Review by abantecart — CHANGES_REQUESTED** [review]

File: public_html/core/engine/dispatcher.php (hunk 4 above)
Comment (abolabo) [reply public_html/core/engine/dispatcher.php:279]: https://www.php.net/manual/en/class.reflectionmethod.php

File: public_html/core/engine/dispatcher.php (hunk 1 above)
Comment (abolabo) [reply public_html/core/engine/dispatcher.php:355]: what type of data we should return if controller type page or unknown?

File: public_html/core/engine/dispatcher.php (hunk 5 above)
Comment (abolabo) [reply public_html/core/engine/dispatcher.php:295]: no sure

File: public_html/system/config.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+define('DB_USERNAME', 'desss');
+define('DB_PASSWORD', 'abolabo');
+define('DB_DATABASE', '1.2.17');
+define('DB_PREFIX', 'cba_');
+
+define('CACHE_DRIVER', 'file');
+// Unique AbanteCart store ID
+define('UNIQUE_ID', 'a4c4d14d0f5d5783af2f631ad13a2e62');
+// Encryption key for protecting sensitive information. NOTE: Change of this key will cause a loss of all existing encrypted information!
+define('ENCRYPTION_KEY', 'sC15Yw');
```
Comment (abantecart) [inline public_html/system/config.php:26]: Fix this and add to gitignore.

**Review by abantecart — CHANGES_REQUESTED** [review]

**Pull request merged**